From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Taiyou06 <kaandindar21@gmail.com>
Date: Wed, 30 Jul 2025 23:18:41 +0200
Subject: [PATCH] Lithium: combined heightmap update

This patch is based on the following mixins:
* "net/caffeinemc/mods/lithium/mixin/world/combined_heightmap_update/LevelChunkMixin.java"
* "net/caffeinemc/mods/lithium/mixin/world/combined_heightmap_update/HeightmapAccessor.java"
* "net/caffeinemc/mods/lithium/common/world/chunk/heightmap/CombinedHeightmapUpdate.java"

By: 2No2Name <2No2Name@web.de>
As part of: Lithium (https://github.com/CaffeineMC/lithium)
Licensed under: LGPL-3.0-only (https://www.gnu.org/licenses/lgpl-3.0.html)

// SYNC SOURCE START
rev: 11ea08adce6f94627d3eaaa65506a60e0451c8b6
branch: develop
repo: https://github.com/CaffeineMC/lithium.git
track:
- "common/src/main/java/net/caffeinemc/mods/lithium/mixin/world/combined_heightmap_update/LevelChunkMixin.java"
- "common/src/main/java/net/caffeinemc/mods/lithium/mixin/world/combined_heightmap_update/HeightmapAccessor.java"
- "common/src/main/java/net/caffeinemc/mods/lithium/common/world/chunk/heightmap/CombinedHeightmapUpdate.java"
// SYNC SOURCE END

diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index a01f41527a5af361efeb136be82667caa3c5b700..349f7fd0058301cf45477a4ce40693a9fd17610f 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -439,10 +439,13 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
         // Leaf end - optimize random tick
 
         Block newBlock = state.getBlock();
-        this.heightmaps.get(Heightmap.Types.MOTION_BLOCKING).update(localX, y, localZ, state);
-        this.heightmaps.get(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES).update(localX, y, localZ, state);
-        this.heightmaps.get(Heightmap.Types.OCEAN_FLOOR).update(localX, y, localZ, state);
-        this.heightmaps.get(Heightmap.Types.WORLD_SURFACE).update(localX, y, localZ, state);
+        // Leaf start - Lithium - combined heightmap update
+        Heightmap heightmap0 = this.heightmaps.get(Heightmap.Types.MOTION_BLOCKING);
+        Heightmap heightmap1 = this.heightmaps.get(Heightmap.Types.MOTION_BLOCKING_NO_LEAVES);
+        Heightmap heightmap2 = this.heightmaps.get(Heightmap.Types.OCEAN_FLOOR);
+        Heightmap heightmap3 = this.heightmaps.get(Heightmap.Types.WORLD_SURFACE);
+        net.caffeinemc.mods.lithium.common.world.chunk.heightmap.CombinedHeightmapUpdate.updateHeightmaps(heightmap0, heightmap1, heightmap2, heightmap3, this, localX, y, localZ, state);
+        // Leaf end - Lithium - combined heightmap update
         boolean isEmpty = section.hasOnlyAir();
         if (wasEmpty != isEmpty) {
             this.level.getChunkSource().getLightEngine().updateSectionStatus(pos, isEmpty);
diff --git a/net/minecraft/world/level/levelgen/Heightmap.java b/net/minecraft/world/level/levelgen/Heightmap.java
index 22990ffcb40304279530462a72c489e8e1207e08..84a1455c153e2ecbfc3e128903dd83ee9a53e952 100644
--- a/net/minecraft/world/level/levelgen/Heightmap.java
+++ b/net/minecraft/world/level/levelgen/Heightmap.java
@@ -49,6 +49,12 @@ public class Heightmap {
         this.data = new SimpleBitStorage(heightBits, 256);
     }
 
+    // Leaf start - Lithium - combined heightmap update
+    public final Predicate<BlockState> isOpaque() {
+        return this.isOpaque;
+    }
+    // Leaf end - Lithium - combined heightmap update
+
     public static void primeHeightmaps(final ChunkAccess chunk, final Set<Heightmap.Types> types) {
         if (!types.isEmpty()) {
             int size = types.size();
@@ -131,7 +137,7 @@ public class Heightmap {
         return (int)this.heightmapRaw[index] + this.chunk.getMinY(); // Paper - optimise heightmap access
     }
 
-    private void setHeight(final int x, final int z, final int height) {
+    public void setHeight(final int x, final int z, final int height) { // Leaf - Lithium - combined heightmap update - private -> public
         // Paper start - optimise heightmap access
         final int index = getIndex(x, z);
         final int value = height - this.chunk.getMinY();
